src/js_realm.hpp

```diff
@@ -874,61 +888,50 @@ void RealmClass<T>::async_open_realm(ContextType ctx, ObjectType this_object, Ar
         Function<T>::callback(protected_ctx, protected_callback, protected_this, 1, callback_arguments);
     }
 
-    // First download the Realm with no schema set to avoid spurious writes that
-    // the server may either reject (for read-only Realms) or just waste time
-    // merging.
-    std::shared_ptr<Realm> realm;
-    try {
-        auto download_config = config;
-        download_config.schema = util::none;
-        download_config.cache = false;
-        realm = realm::Realm::get_shared_realm(std::move(download_config));
-    }
-    catch (const RealmFileException& ex) {
-        handleRealmFileException(ctx, config, ex);
-    }
-
-    auto session = user->session_for_on_disk_path(realm->config().path);
-    EventLoopDispatcher<WaitHandler> wait_handler([=, config=std::move(config),
-                                                   defaults=std::move(defaults),
-                                                   constructors=std::move(constructors)](std::error_code error_code) mutable {
+    std::shared_ptr<AsyncOpenTask>* ptr = new std::shared_ptr<AsyncOpenTask>(Realm::get_synchronized_realm(config));
+    EventLoopDispatcher<RealmCallbackHandler> callback_handler([=, config=std::move(config),
+                                                               defaults=std::move(defaults),
+                                                               constructors=std::move(constructors)](std::shared_ptr<Realm> realm, std::exception_ptr error) mutable {
         HANDLESCOPE
-        if (error_code) {
-            ObjectType object = Object::create_empty(protected_ctx);
-            Object::set_property(protected_ctx, object, "message", Value::from_string(protected_ctx, error_code.message()));
-            Object::set_property(protected_ctx, object, "errorCode", Value::from_number(protected_ctx, error_code.value()));
-
-            ValueType callback_arguments[2];
-            callback_arguments[0] = Value::from_null(protected_ctx);
-            callback_arguments[1] = object;
 
-            Function<T>::callback(protected_ctx, protected_callback, typename T::Object(), 2, callback_arguments);
-            return;
-        }
-
-        // Ensure that all of our metadata tables are properly initialized if
-        // this is a query-based sync Realm
-        if (config.sync_config->is_partial) {
-            realm->update_schema({}, 0);
+        // Since the callback is only ever invoked once, make sure to release reference to AsyncOpenTask, so internal
+        // resources like the RealmCoordinator are correctly released.
+        delete ptr;
```

Comment:
This is invalid. `AsyncOpenTaskClass<T>` will also call `delete` on the pointer when it is destroyed. The correct way to do this is to call `set_internal(nullptr)`.